src/System.Diagnostics.TraceSource/src/System/Diagnostics/CorrelationManager.cs

```diff
@@ -0,0 +1,72 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    public class CorrelationManager
+    {
+        internal CorrelationManager()
+        {
+        }
+
+        private AsyncLocal<Stack> _idStack = new AsyncLocal<Stack>();
+        private AsyncLocal<Guid> _idSlotName = new AsyncLocal<Guid>();
+
+        public Guid ActivityId
+        {
+            get
+            {
+                return _idSlotName.Value;
+            }
+            set
+            {
+                _idSlotName.Value = value;
+            }
+        }
+
+        public Stack LogicalOperationStack
+        {
+            get
+            {
+                return GetLogicalOperationStack();
+            }
+        }
+
+        public void StartLogicalOperation(object operationId)
+        {
+            if (operationId == null)
+            {
+                throw new ArgumentNullException(nameof(operationId));
+            }
+
+            Stack _idStack = GetLogicalOperationStack();
+            _idStack.Push(operationId);
+        }
+
+        public void StartLogicalOperation()
+        {
+            StartLogicalOperation(Guid.NewGuid());
+        }
+
+        public void StopLogicalOperation()
+        {
+            Stack _idStack = GetLogicalOperationStack();
+            _idStack.Pop();
```

Comment:
This will throw if there wasn't a matching start, such that the stack is empty.  I assume that's intentional and matches the desktop behavior?